internal/app/upstream/client.go

```diff
@@ -100,50 +105,78 @@ func New(
 		return nil, err
 	}
 
-	ldsClient := v2.NewListenerDiscoveryServiceClient(conn)
-	rdsClient := v2.NewRouteDiscoveryServiceClient(conn)
-	edsClient := v2.NewEndpointDiscoveryServiceClient(conn)
-	cdsClient := v2.NewClusterDiscoveryServiceClient(conn)
-
-	go shutDown(ctx, conn)
-
-	return &client{
-		ldsClient:   ldsClient,
-		rdsClient:   rdsClient,
-		edsClient:   edsClient,
-		cdsClient:   cdsClient,
+	client := &client{
+		ads:         ads,
 		callOptions: callOptions,
 		logger:      namedLogger,
 		scope:       subScope,
-	}, nil
+	}
+
+	if ads {
+		client.adsClient = discovery.NewAggregatedDiscoveryServiceClient(conn)
+	} else {
+		client.ldsClient = v2.NewListenerDiscoveryServiceClient(conn)
+		client.rdsClient = v2.NewRouteDiscoveryServiceClient(conn)
+		client.edsClient = v2.NewEndpointDiscoveryServiceClient(conn)
+		client.cdsClient = v2.NewClusterDiscoveryServiceClient(conn)
+	}
```

Comment:
> ADS and non-ADS clients both connected to the same relay instance

I think this case is supported, since non-ADS grpc services are registered on bootstrap. Requests from downstream non-ADS clients can be sent to ADS upstream, and seem to be working fine.  This can be verified by running the ads example, updating `dynamic_resources` in `example/ads/front-envoy.yaml` with following config

``` yaml
dynamic_resources:
  cds_config:
    api_config_source:
      api_type: GRPC
      grpc_services:
      - envoy_grpc:
          cluster_name: xds-grpc
      set_node_on_first_message_only: true
  lds_config:
    ads: {}
  ads_config:
    api_type: GRPC
    grpc_services:
    - envoy_grpc:
        cluster_name: xds-grpc
```